### PR TITLE
Add features

### LINKED ISSUE #79: Expose loss and optimization functions in config file
At the moment only one loss and one optimization function are implemented. For a more generic usage, one should be able to choose those.

### LINKED ISSUE #99: Class metric log file name
Full [hypertrainer](https://github.com/lemairecarl/hypertrainer) support requires renaming the log files name, for classwise metrics.  

Naming convention for log files supported by hypertrainer can be found [here](https://github.com/lemairecarl/hypertrainer#log-formats)

### PR DESCRIPTION
**Pull request by remtav**
- expose loss / optimizer
- add models from torchvision
- add parameter 'scale_data' to scale values in images (ex.: from [0,255] to [0,1])
- load pretrained dictionary without classifier layer
- add debug mode (progress bar with stats)
- reorganize root folder (e.g. keep only main files in root, send others to /utils)  

Closes #79  
Closes #99

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -84,10 +86,11 @@ global:
   num_classes: 2                    # Number of classes
   data_path: /path/to/data/folder   # Path to folder containing samples
   number_of_bands: 3                # Number of bands in input images
-  model_name: unetsmall				# One of unet, unetsmall, checkpointed_unet, ternausnet, or inception
+  model_name: unetsmall             # One of unet, unetsmall, checkpointed_unet, ternausnet, fcn_resnet101, deeplabv3_resnet101 or inception
   bucket_name:                      # Name of the S3 bucket where data is stored. Leave blank if using local files
   task: segmentation                # Task to perform. Either segmentation or classification.
   num_gpus: 0                       # Number of GPU device(s) to use. Default: 0
+  debug_mode: True                  # if True, displays useful info for debugging
```
Comment (mpelchat04) [inline README.md]: scale_data: [0, 1]                # Min and Max for input data rescaling. Default: [0, 1]. Enter False if no rescaling is desired.

File: images_to_samples.py [hunk 2]
```diff
@@ -183,54 +187,65 @@ def main(params):
...
+            validate_num_classes(info['gpkg'], params['global']['num_classes'], info['attribute_name'])
+
+            # Burn vector file in a raster file
+            np_label_raster = vector_to_raster(info['gpkg'], info['tif'], info['attribute_name'])
+
+            # Guidelines for pre-processing: http://cs231n.github.io/neural-networks-2/#datapre
+            # Scale arrays to values [0,1]. Default: will scale. Useful if dealing with 8 bit *and* 16 bit images.
+            scale = params['sample']['scale_data'] if params['sample']['scale_data'] else True
+            if scale:
+                min, max = params['sample']['scale_data']
```
Comment (mpelchat04) [inline images_to_samples.py]: min and max also refers to python's built-in functions. Rename for clarification.

File: conf/config.yaml [hunk 3]
```diff
@@ -13,10 +13,11 @@ global:
   num_classes: 2
   data_path: /path/to/data/
   number_of_bands: 3
-  model_name: unetsmall     # One of unet, unetsmall, checkpointed_unet or ternausnet
+  model_name: unetsmall     # One of unet, unetsmall, checkpointed_unet, ternausnet, fcn_resnet101, deeplabv3_resnet50,deeplabv3_resnet101
   bucket_name:   # name of the S3 bucket where data is stored. Leave blank if using local files
   task: segmentation  # Task to perform. Either segmentation or classification
   num_gpus: 2
+  debug_mode: True
```
Comment (mpelchat04) [inline conf/config.yaml]: Add `scale_data: [0, 1]`

File: conf/config_ci_classification_local.yaml [hunk 4]
```diff
@@ -25,6 +26,7 @@ sample:
   samples_dist: 200
   min_annotated_percent: 10 # Min % of non background pixels in stored samples. Default: 0
   mask_reference: False
+  scale_data: [0,1]
```
Comment (mpelchat04) [inline conf/config_ci_classification_local.yaml]: `scale_data` should be a global parameter, since it has to be used for `images_to_samples.py` and `inference.py`.

File: conf/config_ci_classification_local.yaml [hunk 5]
```diff
@@ -65,5 +69,9 @@ models:
     pretrained: ./models/TernausNet.pt    # Mandatory
   checkpointed_unet: 
     <<: *unet001
+  fcn_resnet101:   # only for 3 band data
+    pretrained:    # optional
+  deeplabv3_resnet101:   # only for 3 band data
+    pretrained: /home/rtavon/Documents/kingston-test-deeplabv3-2/model/checkpoint.pth.tar   # optional
```
Comment (mpelchat04) [inline conf/config_ci_classification_local.yaml]: Remove personnal pretrained path.

File: losses/__init__.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+        super().__init__()
+        if loss_type == 'CrossEntropy':
+            self.criterion = nn.CrossEntropyLoss(**kwargs)
+        elif loss_type == 'Focal':
+            self.criterion = FocalLoss(**kwargs)
+        elif loss_type == 'Lovasz':
+            self.criterion = LovaszSoftmax(**kwargs)
+        elif loss_type == 'OhemCrossEntropy':
+            self.criterion = OhemCrossEntropy2d(**kwargs)
+        elif loss_type == 'Focal':
```
Comment (mpelchat04) [inline losses/__init__.py]: `Focal` loss mentioned twice.

File: conf/config.yaml [hunk 7]
```diff
@@ -35,6 +36,8 @@ training:
   num_tst_samples: 1000
   batch_size: 32
   num_epochs: 100
+  loss_fn: Lovasz # One of CrossEntropy or Lovasz (*Lovasz for segmentation tasks only)
```
Comment (mpelchat04) [inline conf/config.yaml]: `OhemCrossEntropy2d` is also valid

File: models/TernausNet.py [hunk 8]
```diff
@@ -10,7 +10,7 @@
 # import torch should be first. Unclear issue, mentioned here: https://github.com/pytorch/pytorch/issues/2083
 from torch import nn
 from torchvision import models
-from utils import load_from_checkpoint
+from utils.utils import load_from_checkpoint, chop_layer
```
Comment (mpelchat04) [inline models/TernausNet.py]: `chop_layer` unused import.

File: models/unet.py [hunk 9]
```diff
@@ -1,4 +1,5 @@
 import torch, utils
```
Comment (mpelchat04) [inline models/unet.py]: first `utils` import irrelevant

File: utils/optimizer.py [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+
+def create_optimizer(params, mode='adam', base_lr=1e-3, weight_decay=4e-5):
+    if mode == 'adam':
+        optimizer = optim.Adam(params, lr=base_lr)
+    elif mode == 'sgd':
+        optimizer = optim.SGD(params, lr=base_lr, momentum=0.9, weight_decay=weight_decay)
+    elif mode == 'adabound':
+        optimizer = AdaBound(params, lr=base_lr, final_lr=base_lr*100, gamma=1e-3, eps=1e-8, weight_decay=weight_decay)
+    else:
```
Comment (mpelchat04) [inline utils/optimizer.py]: `AdaBoundW` is also available in `adabound.py`. Could it be included as an optimizer choice?
